### ISSUE TITLE
Incompatible return value type when a function returns a Callable

### ISSUE THREAD

**Issue by andymwood** [python/mypy#16290]
**Bug Report**

Mypy reports an incompatible return value on the following code, but I believe it shouldn't.

**To Reproduce**

```python
from typing import Callable
from mypy_extensions import KwArg


def k() -> Callable[[bool, KwArg(int)], int]:
    def func(flag: bool, **kw: int) -> int:
        return 0
    return func
```

**Expected Behavior**
```console
$ mypy test.py 
Success: no issues found in 1 source file
```

**Actual Behavior**
```console
$ mypy test.py 
test.py:8: error: Incompatible return value type (got "Callable[[Arg(bool, 'flag'), KwArg(int)], int]", expected "Callable[[bool, KwArg(int)], int]")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

The problem doesn't occur if `VarArg` and `*args` are used instead of `KwArg` and `**kw`.

**Your Environment**

- Mypy version used: 1.6.1
- Python version used: 3.10.13

**Comment by AlexWaygood** [comment]
I'm not sure if this is expected behaviour or not. But I'm afraid these mypy-specific extensions to the typing system are documented as deprecated, so I don't think we'll be fixing this: https://mypy.readthedocs.io/en/stable/additional_features.html#extended-callable-types

**Comment by erictraut** [comment]
The problem is that your inner `func` does not declare the `flag` parameter as positional-only, but the return type to `k` requires that it is positional-only. If you change the name of `flag` to `__flag` or insert a positional-only separator, it will type check without error.

```python
    def func(flag: bool, /, **kw: int) -> int:
        return 0
```

Alternatively, you can replace the mypy-specific extensions with a standard callback protocol.

```python
from typing import Protocol

class Callback(Protocol):
    def __call__(self, flag: bool, **kw: int) -> int:
        ...

def k() -> Callback:
    def func(flag: bool, **kw: int) -> int:
        return 0

    return func
```
